src/mame/misc/vocalizer.cpp

```diff
@@ -0,0 +1,565 @@
+	auto rx_irq_cb() { return m_rx_irq_cb.bind(); }
+
+	u8 read() { return get_received_char(); }
+	void write(u8 data) { transmit_register_setup(data); }
+
+protected:
+	virtual void device_start() override {};
+	virtual void device_reset() override;
+
+	virtual void tra_callback() override { m_tx_cb(transmit_register_get_data_bit()); }
+	virtual void tra_complete() override { m_tx_irq_cb(1); }
+	virtual void rcv_complete() override { receive_register_extract(); m_rx_irq_cb(1); }
+
+	devcb_write_line m_tx_cb;
+	devcb_write_line m_tx_irq_cb;
+	devcb_write_line m_rx_irq_cb;
+};
+
+DEFINE_DEVICE_TYPE(VOCALIZER_UART, vocalizer_uart_device, "vocalizer_uart", "Vocalizer 1000 UART")
+
+vocalizer_uart_device::vocalizer_uart_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock):
+	device_t(mconfig, VOCALIZER_UART, tag, owner, clock),
+	device_serial_interface(mconfig, *this),
+	m_tx_cb(*this),
+	m_tx_irq_cb(*this),
+	m_rx_irq_cb(*this)
+{
+}
+
+void vocalizer_uart_device::device_reset()
+{
+	set_data_frame(1, 8, PARITY_NONE, STOP_BITS_1);
+	set_rate(31250);
+}
+
+namespace {
+
+//**************************************************************************
+class vocalizer_state : public driver_device
+{
+public:
+	vocalizer_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_lcdc(*this, "lcdc"),
+		m_es5503(*this, "es5503"),
+		m_uart(*this, "uart%u", 0),
+		m_bank(*this, "rombank"),
+		m_keys(*this, "IN%u", 0)
+	{ }
+
+	void vocalizer(machine_config &config);
+
+	void vocalizer_palette(palette_device &palette) const;
+	HD44780_PIXEL_UPDATE(lcd_pixel_update);
+
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+private:
+	enum
+	{
+		FIRQ_TX0 = (1<<0), // UART 0 Tx (link cable)
+		FIRQ_RX  = (1<<1), // UART Rx (either one)
+		FIRQ_MIC = (1<<2), // audio input detection (triggered by input waveform edge, or something else?)
+		FIRQ_TX1 = (1<<3)  // UART 1 Tx (MIDI)
+	};
+
+	void maincpu_map(address_map &map);
+	void sound_map(address_map &map);
+
+	template<int Shift>	u8 time_r();
+
+	u8 uart_rx_r();
+	template<int Num> void uart_rx_irq(int state);
+
+	u8 status_r();
+	void config_w(u8 data) { m_config = data; }
+	void input_sel_w(u8 data) { m_input_sel = BIT(data, 3, 4); }
+
+	template<int Num> void firq_w(int state);
+	void firq_ack_w(u8 data);
+	void firq_mask_w(u8 data);
+	void firq_update();
+
+	void volume_w(u8 data);
+
+	void apo_w(u8 data);
+
+	template<int Num> void bank_w(u8 data) { m_bank->set_entry(Num); }
+
+	required_device<mc6809e_device> m_maincpu;
+	required_device<hd44780_device> m_lcdc;
+	required_device<es5503_device> m_es5503;
+	required_device_array<vocalizer_uart_device, 2> m_uart;
+	required_memory_bank m_bank;
+	required_ioport_array<12> m_keys;
+
+	u8 m_power;
+	u8 m_config;
+	u8 m_input_sel;
+	u8 m_firq_status, m_firq_mask;
```

Comment:
FYI initial variables are uncertain at MAME start if you don't initialize them (here, or at vocalizer_state(x..) or at machine_start.

Reply:
They're initialized in machine_reset, like I've normally been doing in other drivers.

Reply:
Ah I missed that. It's probably not the best place when you consider that devices do device_reset before the driver does machine_reset. If a device accesses a devcb at reset that goes back to the driver, it may read an uninitialized var, or write to one that gets overwritten at machine_reset (some devices initialize output ports at reset).